R/run_sdmtmb.R

```diff
@@ -0,0 +1,224 @@
+
+		if (obs == "gamma"){
+			family = sdmTMB::delta_gamma()
+		} 
+		if (obs == "lognormal"){
+			family = sdmTMB::delta_lognormal()
+		} 
+		if (!obs %in% c("gamma", "lognormal")){
+			# Add error statement and stop run
+		}
+		
+		# Create mesh
+		mesh_survey <- VASTWestCoast::convert_survey4vast(
+			survey = surveyspp["survey"])
+
+		mesh <- VASTWestCoast::VAST_mesh(
+			data = catch_data,
+		  survey = mesh_survey,
+		  numknots = knots,
+		  savedir = file.path(sppdir, survey, "index", obs))
+		data_mesh <- mesh$mesh$data.inner
+				
+		# Create prediction grid
+		grid <- VASTWestCoast::get_inputgrid(survey = mesh_survey)
+		grid <- grid[grid$Area_km2 > 0, ]
+		grid <- sdmTMB::add_utm_columns(grid, c("Lon", "Lat"), utm_crs = 32610) # UTM 10
+		grid$pass_scaled <- 0
+		grid$vessel_scaled <- data_mesh$vessel_scaled[1]
+		sub_grid <- dplyr::select(grid, X, Y, Area_km2, pass_scaled, vessel_scaled, Lat, Lon)
+		
+		year_grid <- purrr::map_dfr(data_years, function(yr) {
+		  sub_grid$Year <- yr
+		  sub_grid
+		})
+
+		mesh <- sdmTMB::make_mesh(
+    	data = data_mesh,
+    	xy_cols = c("X", "Y"),
+    	n_knots = knots
+	  )
+
+		# plot and save the mesh
+		grDevices::png(filename = file.path(sppdir, survey, "index", obs, "mesh.png"),
+      width = 7, height = 7, units = "in", res = 300, pointsize = 12)
+    plot(mesh)
+    dev.off()
+
+		data_mesh$vessel_scaled <- as.factor(data_mesh$vessel_scaled)
+		
+		# Run model
+		fit <- sdmTMB::sdmTMB(
+  		formula = formula,
+  		time = "Year",
+  		offset = log(data_mesh$Area_swept_km2),
+  		data = data_mesh,
+  		mesh = mesh,
+  		family = family,
+  		spatial = "on",
+  		spatiotemporal = list("iid", "iid"),
+  		anisotropy = TRUE,
+  		silent = TRUE,
+  		control = sdmTMB::sdmTMBcontrol(
+  		  newton_loops = 1L,
+  		  map = list(ln_H_input = factor(c(1, 2, 1, 2))) # <- force sdmTMB to share anisotropy parameters across the two delta models
+  		),
+  		do_index = TRUE,
+  		predict_args = list(newdata = year_grid, re_form_iid = NA),
+  		index_args = list(area = year_grid$Area_km2)
+		)
+
+		index <- sdmTMB::get_index(
+				  fit, 
+				  bias_correct = FALSE # TRUE
+		)
+		index$area <- "coastwide" 
+
+		get_diagnostics(
+			dir = file.path(sppdir, survey, "index", obs), 
+			fit = fit, 
+			prediction_grid = year_grid
+		)
+
+		# Create indices by area
+		area_indices <- NULL 
+		for (strata in strata.limits[-1,"STRATA"]){
+
+			sub_year_grid <- year_grid %>% 
+				dplyr::filter(Lat < strata.limits[strata.limits$STRAT == strata, "north_border"]) %>%
+				dplyr::filter(Lat >= strata.limits[strata.limits$STRAT == strata, "south_border"])
+			
+			pred_area <- predict( #does not work with sdmTMB::predict
+				fit, 
+				newdata = sub_year_grid,
+			  return_tmb_object = TRUE
+			)
+
+			sub_index <- sdmTMB::get_index(
+		  	pred_area, #fit, # skipping prediction step
+		  	area = sub_year_grid$Area_km2,
+		  	bias_correct = FALSE # TRUE
```

Comment:
Here and elsewhere, I imagine you'll be swapping this to `TRUE`. It's not nearly as computationally expensive as it once was (one `MakeADFun` and one gradient evaluation).